### ISSUE TITLE
KO - SERVER DOWN

### ISSUE THREAD

**Issue by MrPowerGamerBR** [markozajc/Akiwrapper#14]
Looks like the English and Portuguese servers are down... or maybe the server checker is broken?

```
java.lang.reflect.InvocationTargetException: null
 at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
 at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
 at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 at java.base/java.lang.reflect.Method.invoke(Method.java:567)
 at kotlin.reflect.jvm.internal.calls.CallerImpl$Method.callMethod(CallerImpl.kt:97)
 at kotlin.reflect.jvm.internal.calls.CallerImpl$Method$Instance.call(CallerImpl.kt:113)
 at kotlin.reflect.jvm.internal.KCallableImpl.call(KCallableImpl.kt:106)
 at kotlin.reflect.jvm.internal.KCallableImpl.callDefaultMethod$kotlin_reflection(KCallableImpl.kt:152)
 at kotlin.reflect.full.KCallables.callSuspendBy(KCallables.kt:73)
 at net.perfectdreams.commands.manager.CommandManager.execute(CommandManager.kt:255)
 at net.perfectdreams.loritta.platform.discord.commands.DiscordCommandManager.dispatch(DiscordCommandManager.kt:484)
 at net.perfectdreams.loritta.platform.discord.commands.DiscordCommandManager$dispatch$3.invokeSuspend(DiscordCommandManager.kt)
 at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
 at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
 at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
 at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
 at java.base/java.lang.Thread.run(Thread.java:835)
Caused by: com.markozajc.akiwrapper.core.exceptions.ServerGroupUnavailableException: KO - SERVER DOWN
 at com.markozajc.akiwrapper.core.utils.Servers.getFirstAvailableServer(Servers.java:178)
 at com.markozajc.akiwrapper.core.impl.AkiwrapperImpl.<init>(AkiwrapperImpl.java:110)
 at com.markozajc.akiwrapper.AkiwrapperBuilder.build(AkiwrapperBuilder.java:72)
 at net.perfectdreams.loritta.commands.vanilla.fun.AkinatorCommand.run(AkinatorCommand.kt:202)
 ... 17 common frames omitted
```

**Comment by markozajc** [comment]
Looks like they've implemented scopes (person, object, animal) to English servers (pretty sure they were only available on French and some other languages). This means that I will likely need to update the server scanner. You could _probably_ temporarily fix it by generating a server list yourself with the current AkinatorServerScanner (or even update it, PRs are welcome), but that could mean that a random scope will be used every time.

**Comment by markozajc** [comment]
Forget what I said, they've relocated a bunch of their APIs. Looks like the API servers are no longer accessed directly, but are proxied through the akinator.com website. It also appears that they still have to be known, and are passed through the `urlApiWs` querystring parameter. Reverse engineering the changes and patching Akiwrapper might thus take some time.

**Comment by ahmed605** [comment]
@markozajc I have found a workaround and I pushed it into an another C# project (library), so if you want to take a look at it here's the commit link: https://github.com/janniksam/Akinator.Api.Net/commit/766d0a3faefb3f34a1cad80ed53bfd0360e74df9

**Comment by markozajc** [comment]
@ahmed6p5 Thanks! Could you briefly summarize the API changes, as understanding exactly what each change does would require me to understand the whole library as well, which I don't. In general, I want to know
* whether the srv#.akinator.com servers are still accessible directly
* whether the headers you've added in the patch are mandatory
Thanks in advance.

**Comment by ahmed605** [comment]
@markozajc 

- srv#.akinator.com is accessible in all API calls except the call that creates a new session

- Yeah, these headers are mandatory (I have tested removing every header of them sperately but it didn't work so all of them are mandatory)

**Comment by ahmed605** [comment]
@markozajc also, take a look at these:

- https://github.com/janniksam/Akinator.Api.Net/issues/11#issuecomment-608591987

- https://github.com/janniksam/Akinator.Api.Net/issues/11#issuecomment-612018632

**Comment by ahmed605** [comment]
@markozajc if you need a brief better description/explanation of the API changes just tell/ping me :)

**Comment by markozajc** [comment]
@ahmed605 Thanks for the server catalogue API, it'll make the transition much easier! I'd just like to know what the `footprint` querystring parameter does and how it can be obtained (since it seems to require a specific value).
Thanks in advance.

**Comment by ahmed605** [comment]
@markozajc I don't know how it generated I have found that Value and API endpoint while reversing engineering the Android app and it was typed as it is in code (not generated)

**Comment by ahmed605** [comment]
@markozajc Also as you can see in the commit you should disable cookies besides adding headers

**Comment by ahmed605** [comment]
@markozajc BTW have found an endpoint that returns a list of all languages, servers, supported server types and URLs `https://global3.akinator.com/ws/instances_v2.php?media_id=14&footprint=cd8e6509f3420878e18d75b9831b317f&mode=https`
![decompiled code](https://user-images.githubusercontent.com/34550324/79042707-24f07400-7bfa-11ea-87cb-45f592b9de1d.PNG)

**Comment by LagBug** [comment]
Hi, do we have an ETA on when this could be fixed?

**Comment by markozajc** [comment]
@LagBug Currently not, and I'm also quite busy working on some other projects. PRs are welcome.

**Comment by LagBug** [comment]
All good, take your time. I personally am really busy as well, so I can't contribute to Akiwrapper at this time.

**Comment by MrPowerGamerBR** [comment]
I'm tired of hearing people saying "why is +akinator broken???" so I will give 5 USD to the person that fixes this issue.

If I decide to withdraw this bounty (if I decide to just scrap the command altogether or fix it myself) I will edit this comment.

**Comment by markozajc** [comment]
Finally have some spare time on my hands, so I've decided to give this issue a go. Progress is being made on the api_update branch. This will likely require me to rewrite a large portion of the library due to the way things are put together, so it will probably be a breaking update.

**Comment by markozajc** [comment]
Nevermind, false alarm, I'm still busy as fuck.

**Comment by markozajc** [comment]
Fixed in 1.5
